graylog2-server/src/main/java/org/graylog/plugins/views/startpage/StartPageService.java

```diff
@@ -68,26 +68,32 @@ public PaginatedResponse<LastOpened> findLastOpenedFor(final SearchUser searchUs
                 .items()
                 .stream()
                 .skip((long) (page - 1) * perPage)
+                .map(i -> startPageItemTitleRetriever
+                        .retrieveTitle(i.grn())
+                        .map(title -> new LastOpened(i.grn(), title, i.timestamp()))
+                        .orElse(null))
+                .filter(Objects::nonNull)
                 .limit(perPage)
-                .map(i -> new LastOpened(i.grn(), catalog.getTitle(i.grn()), i.timestamp()))
                 .toList();
 
         return PaginatedResponse.create("lastOpened", new PaginatedList<>(items, items.size(), page, perPage));
     }
 
-    private String getTitle(RecentActivityDTO i) {
-        return i.itemTitle() == null ? catalog.getTitle(i.itemGrn()) : i.itemTitle();
-    }
-
     public PaginatedResponse<RecentActivity> findRecentActivityFor(final SearchUser searchUser, int page, int perPage) {
         final var items = recentActivityService.findRecentActivitiesFor(searchUser, page, perPage);
         final var mapped = items.stream()
-                 .map(i -> new RecentActivity(i.id(),
-                        i.activityType(),
-                        i.itemGrn(),
-                        getTitle(i),
-                        i.userName(),
-                        i.timestamp())).toList();
+                .map(i -> startPageItemTitleRetriever
```

Comment:
You can also use `Stream#flatMap` here to get rid of the `orElse`/`filter`-combination below.
```suggestion
                .flatMap(i -> startPageItemTitleRetriever
```

Reply:
Don't you have a feeling that it looks tricky and my explicit approach is more clear?

Reply:
No, not at all. This is exactly what `Optional#stream` is made for and IMO it is clearer than unwrapping an `Optional` with `null` (especially considering that `Optional`s were meant to avoid `null`-handling like this).